src/common/sensor.rs

```diff
@@ -35,7 +35,7 @@ pub trait Sensor: Readings + Status + DoCommand {}
 pub type SensorType = Arc<Mutex<dyn Sensor>>;
 
 pub trait SensorT<T>: Sensor {
-    fn get_readings(&self) -> anyhow::Result<TypedReadingsResult<T>>;
+    fn get_readings(&mut self) -> anyhow::Result<TypedReadingsResult<T>>;
```

Comment:
I see why we need this for ultrasonic sensor but I would like to avoid changing the signature to a &mut self if possible. Would we be able to get around this limitation with interior mutability?

Reply:
Sure, I'll give interior mutability a try. I wasn't exactly thrilled about needing to switch it to `&mut self` either.

Note though that `Readings::get_generic_readings` is `&mut self` too as of #126. If it is a design goal for the sensor API to avoid mutability for readings, perhaps that needs to be revisited?

Reply:
Updated with interior mutability by wrapping the `PinDriver`s with `RefCell` and sprinkling `borrow_mut` as needed. I'm not sure if this is better or not. The sensor, for better or worse, really is changing the state of the outside world by driving pins and configuring interrupts. Requiring it to remain `&self` means any sort of stateful sensor will have little choice but to similarly adopt interior mutability. On the other hand, maybe the majority of sensors are passive. I think it comes down to what we want in terms of API design and implementation experience for module authors and how that intersects with idiomatic rust API design, neither of which is something I'm particularly expert in.

Reply:
I not an expert either with idiomatic Rust API design but I feels more natural to have get_reading be non mutable and have the sensor update it's state if need in a separate thread/future or worst case via interior mutability (in that case concurrent reading should yield a previous value) With respect to  `Readings::get_generic_readings` i forgot why we ended up with mut self here since it delegate all call to `get_readings`